File: Step4/Step4_Simulation.f90
program step4_comparison_sweep
    ! ============================================
    ! Step 4: V-PCM vs NPU 比較シミュレーション (Fortran版・新計画)
    ! 目的：多様体スケーリング効率とノイズ耐性の二重検証
    ! ============================================
    implicit none
    integer, parameter :: MAX_N = 512
    integer, dimension(5) :: sizes = (/64, 128, 256, 384, 512/)
    real, dimension(MAX_N, MAX_N) :: A, B, C
    real :: start_t, end_t, noise_lvl
    real :: stability_vpcm, stability_npu
    integer :: n_idx, i, j, k, N, steps
    
    print *, "Starting Step 4 Comparison Simulation (Fortran Implementation)..."
    
    ! 1. Task A: Scaling Efficiency (擬似的な計算負荷計測)
    print *, "Task A: Scaling Efficiency Analysis"
    do n_idx = 1, 5
        N = sizes(n_idx)
        call cpu_time(start_t)
        ! PKGF ステップの擬似実行 (行列演算)
        do steps = 1, 10
            A = 0.5
            B = 0.2
            C = matmul(A(1:N,1:N), B(1:N,1:N)) - matmul(B(1:N,1:N), A(1:N,1:N))
        end do
        call cpu_time(end_t)
        print '(A,I5,A,F10.6,A)', " Dim: ", N, " | Latency: ", (end_t - start_t)/10.0, " s"
    end do
    
    ! 2. Task C: Noise Robustness (理論的減衰モデル)
    print *, "Task C: Noise Robustness Analysis"
    do i = 0, 5
        noise_lvl = real(i) * 0.1
        ! V-PCM (PKGF): ノイズをゆらぎとして統合するモデル (Axiom U1/U2)
        stability_vpcm = 1.0 / (1.0 + (noise_lvl**2) * 0.5)
        if (stability_vpcm < 0.7) stability_vpcm = 0.7 ! 物理的底打ち
        
        ! NPU: デジタル誤差による急激な精度低下
        stability_npu = 1.0 / (1.0 + noise_lvl * 5.0)
        
        print '(A,F4.1,A,F6.4,A,F6.4)', " Noise: ", noise_lvl, " | V-PCM: ", stability_vpcm, " | NPU: ", stability_npu
    end do
    
    print *, "Step 4 Fortran Simulation complete."

end program step4_comparison_sweep

File: Step4/Step4_Simulation.py
import numpy as np
import time
import json
import matplotlib.pyplot as plt
from scipy.linalg import svd

# ============================================
# Step 4: V-PCM vs NPU 比較シミュレーション (M1 最適化・新計画版)
# 目的：多様体スケーリング効率 (Axiom A1) と 物理的ゆらぎ耐性 (U1/U2) の検証
# ============================================

def get_effective_rank(K):
    """SVDに基づく動的次元 (d_eff) の算出"""
    try:
        s = svd(K, compute_uv=False)
        p = (s**2) / (np.sum(s**2) + 1e-12)
        p = p[p > 0]
        return np.exp(-np.sum(p * np.log(p)))
    except:
        return 1.0

def simulate_scaling_task():
    """Task A: 多様体スケーリング効率のシミュレート"""
    print("Running Task A: Manifold Scaling Comparison...")
    sizes = [64, 128, 256, 512, 1024]
    times_vpcm = []
    times_npu = []
    
    for N in sizes:
        # V-PCM (GPU/Parallel Proxy): 物理並列性により N に対して O(N log N) または O(N) に近いスケーリングを目指す
        # ここでは行列演算の効率性を擬似的に評価
        A = np.random.rand(N, N).astype(np.float32)
        B = np.random.rand(N, N).astype(np.float32)
        
        start = time.time()
        # PKGF 10ステップ分のシミュレート
        for _ in range(10):
            C = np.dot(A, B) - np.dot(B, A)
            A = A + 0.01 * C
        times_vpcm.append((time.time() - start) / 10.0)
        
        # NPU (Sequential/Tile-based Inference Proxy): 固定サイズを超えると O(N^2.8) 的な増大
        # 従来の行列積のスケーリングを模倣
        start = time.time()
        for _ in range(10):
            _ = np.dot(A, B)
        # NPUのオーバーヘッドをシミュレート
        overhead = (N / 512)**2.5 * 0.001
        times_npu.append(((time.time() - start) / 10.0) + overhead)
        
    return sizes, times_vpcm, times_npu

def simulate_noise_robustness():
    """Task C: 物理的ゆらぎへの耐性比較"""
    print("Running Task C: Noise Robustness Analysis...")
    noise_levels = np.linspace(0, 0.5, 20)
    stability_vpcm = []
    stability_npu = []
    
    N = 128
    base_K = np.eye(N) + np.random.normal(0, 0.1, (N, N))
    
    for noise in noise_levels:
        # V-PCM: Axiom U1/U2。ノイズを「複素ゆらぎ」として統合。
        # ランクの保持能力が高い
        K_noisy = base_K + np.random.normal(0, noise, (N, N))
        # PKGF 統一方程式による再構成（簡易版）
        K_stable = K_noisy / (np.linalg.norm(K_noisy) + 1e-12)
        rank_v = get_effective_rank(K_stable)
        # 安定性スコア: 初期ランクに対する維持率 (0.0 - 1.0)
        stability_vpcm.append(min(rank_v / get_effective_rank(base_K), 1.0))
        
        # NPU: ノイズは単なるビット誤差。構造を線形に破壊。
        # 信号対雑音比 (SNR) に依存
        npu_score = 1.0 / (1.0 + noise * 10.0)
        stability_npu.append(npu_score)
        
    return noise_levels, stability_vpcm, stability_npu

def run_step4_simulation():
    # 1. スケーリング実験
    sizes, t_vpcm, t_npu = simulate_scaling_task()
    
    # 2. ノイズ耐性実験
    noises, s_vpcm, s_npu = simulate_noise_robustness()
    
    # 結果の統合
    results = {
        "scaling": {
            "dimensions": sizes,
            "vpcm_latency": t_vpcm,
            "npu_latency": t_npu
        },
        "robustness": {
            "noise_levels": noises.tolist(),
            "vpcm_stability": s_vpcm,
            "npu_stability": s_npu
        }
    }
    
    with open("Step4/simulation_results.json", "w") as f:
        json.dump(results, f, indent=4)
        
    # 可視化
    plt.figure(figsize=(12, 5))
    
    # Scaling Plot
    plt.subplot(1,2,1)
    plt.plot(sizes, t_vpcm, 'o-', label="V-PCM (PKGF Manifold Flow)", color='red')
    plt.plot(sizes, t_npu, 's--', label="Standard NPU Inference", color='blue')
    plt.yscale('log')
    plt.xscale('log')
    plt.title("Manifold Scaling Efficiency (Axiom A1)")
    plt.xlabel("State Space Dimension (N)")
    plt.ylabel("Latency per Step (s)")
    plt.grid(True, which="both", ls="-", alpha=0.3)
    plt.legend()
    
    # Robustness Plot
    plt.subplot(1,2,2)
    plt.plot(noises, s_vpcm, label="V-PCM (Axiom U1/U2 Integration)", color='red')
    plt.plot(noises, s_npu, label="Standard NPU (Linear Decay)", color='blue')
    plt.title("Structural Robustness vs Noise")
    plt.xlabel("Fluctuation Level (K_fluct)")
    plt.ylabel("Stability Score")
    plt.grid(True, alpha=0.3)
    plt.legend()
    
    plt.tight_layout()
    plt.savefig("Step4/step4_result.png")
    print("Step 4 Simulation complete. Results saved to Step4/simulation_results.json and step4_result.png")

if __name__ == "__main__":
    run_step4_simulation()
